src/vm/peimagelayout.cpp

```diff
@@ -223,6 +218,34 @@ void PEImageLayout::ApplyBaseRelocations()
             }
         }
 
+        BOOL bExecRegion = (dwOldProtection & (PAGE_EXECUTE | PAGE_EXECUTE_READ |
+            PAGE_EXECUTE_READWRITE | PAGE_EXECUTE_WRITECOPY)) != 0;
+
+        // We need to flush instruction cache only for pages that have relocations instead of full sections
+        // because otherwise application's shared clean memory is increased in some cases on Linux.
+        if (bRelocDone && bExecRegion)
+        {
+            SIZE_T cbPageCount = bCrossPageBoundary ? 2 : 1;
+
+            if (pageAddress - pFlushRegion == cbFlushRegion)
+            {
+                cbFlushRegion += cbPageSize * cbPageCount;
+            }
+            else if (pageAddress - pFlushRegion > (ptrdiff_t)cbFlushRegion)
+            {
+                if (pFlushRegion != NULL)
+                {
+                    ClrFlushInstructionCache(pFlushRegion, cbFlushRegion);
+                }
+
+                pFlushRegion = pageAddress;
+                cbFlushRegion = cbPageSize * cbPageCount;
```

Comment:
This still does not look right. You are taking `cbPageCount` that was computed using `bCrossPageBoundary` for previous page with relocs, and using it to start a region for the current page with relocs.